contracts/ssnlist.scilla

```diff
@@ -1542,3 +1571,578 @@ transition DrainContractBalance(amt: Uint128, initiator: ByStr20)
+        | Some existing_lbdc =>
+          (* new deleg has delegate same ssn as requestor *)
+          (* don't replace; already assert no buffered deposit and rewards for both requestor and new_deleg  *)
+          delete last_buf_deposit_cycle_deleg[initial_deleg][ssn_address];
+          e = { _eventname: "SwapLastBuffDepositCycleDelegExists"; initial_deleg: initial_deleg; new_deleg: new_deleg_addr; ssn_addr: ssn_address; existing_lbdc: existing_lbdc };
+          event e
+        | None =>
+          (* new deleg has no last buff deposit for this ssn *)
+          last_buf_deposit_cycle_deleg[new_deleg_addr][ssn_address] := requestor_lbdc;
+          delete last_buf_deposit_cycle_deleg[initial_deleg][ssn_address];
+          e = { _eventname: "SwapLastBuffDepositCycleDelegNewEntry"; initial_deleg: initial_deleg; new_deleg: new_deleg_addr; ssn_addr: ssn_address; lbdc: requestor_lbdc };
+          event e
+        end
+      | None => (* no new deleg to transfer to *)
+      end
+    | None => (* requestor has no last buff deposit cycle at this ssn *)
+    end
+  | None => (* not possible to have no existing deleg *)
+  end
+end
+
+(* @dev Swap deleg address from current_deleg -> new_deleg in withdrawal_pending *)
+procedure SwapPendingWithdrawal(blk_deposit_cycle: Pair BNum Uint128)
+  current_deleg_o <- current_deleg;
+  new_deleg_o <- new_deleg;
+  fst = @fst BNum Uint128;
+  snd = @snd BNum Uint128;
+  blk_num = fst blk_deposit_cycle;
+  deposit = snd blk_deposit_cycle;
+  
+  match new_deleg_o with
+  | Some new_deleg_addr =>
+    new_deleg_pending_withdrawal <- withdrawal_pending[new_deleg_addr][blk_num];
+    match new_deleg_pending_withdrawal with
+    | Some existing_deposit =>
+      (* new deleg has exisiting blk_num *)
+      new_deposit = builtin add existing_deposit deposit;
+      withdrawal_pending[new_deleg_addr][blk_num] := new_deposit;
+      e = { _eventname: "SwapPendingWithdrawalExists"; initial_deleg: current_deleg_o; new_deleg: new_deleg_addr; existing_deposit: existing_deposit; new_deposit: new_deposit };
+      event e
+    | None =>
+      (* new deleg has no blk_num entry*)
+      withdrawal_pending[new_deleg_addr][blk_num] := deposit;
+      e = { _eventname: "SwapPendingWithdrawalNewEntry"; initial_deleg: current_deleg_o; new_deleg: new_deleg_addr; deposit: deposit };
+      event e
+    end
+  | None => (* no new deleg to transfer to *)
+  end
+end
+
+(* @dev Clears the mutable addresses after use *)
+procedure SwapCleanUp()
+  r = Some {ByStr20} reset_addr;
+  current_deleg := r;
+  current_ssn := r;
+  new_deleg := r
+end
+
+
+(* @dev Change a delegator's stake from initiator -> new_deleg_addr. *)
+procedure SwapDelegator(requestor: ByStr20, new_deleg_addr: ByStr20)
+  c = Some {ByStr20} requestor;
+  current_deleg := c;
+  n = Some {ByStr20} new_deleg_addr;
+  new_deleg := n;
+  ssnlist_o <- ssnlist;
+  ssninfo_list = builtin to_list ssnlist_o;
+
+  (* swap *)
+  forall ssninfo_list SwapBuffDeposit;
```

Comment:
I think this is no longer required

Reply:
I agree; can be removed since we assert delegator A has no buffered deposits. `buff_deposit_deleg[A]` should be empty.